fossunited/chapters/doctype/foss_chapter/foss_chapter.py

```diff
@@ -57,13 +57,73 @@ def before_save(self):
 
     def on_update(self):
         self.set_member_roles()
+        self.handle_member_removal()
 
     def set_member_roles(self):
-        for member in self.chapter_members:
-            user = frappe.get_doc("User", member.email)
-            user.add_roles("Chapter Team Member")
-            if member.role == "Lead":
-                user.add_roles("Chapter Lead")
+        current_user = frappe.session.user
+        current_session_data = frappe.session.data
+        frappe.set_user("Administrator")
+
+        try:
+            for member in self.chapter_members:
+                user = frappe.db.get_value(
+                    USER_PROFILE, member.chapter_member, "user"
+                )
+
+                if frappe.db.exists(
+                    "Has Role",
+                    {"role": "Chapter Team Member", "parent": user},
+                ):
+                    continue
+                user_doc = frappe.get_doc("User", user)
+
+                user_doc.add_roles("Chapter Team Member")
+                if member.role == "Lead":
+                    user_doc.add_roles("Chapter Lead")
+
+        except Exception as e:
+            frappe.throw(f"Error: {e}")
+        finally:
+            frappe.set_user(current_user)
+            frappe.session.data = current_session_data
+
+    def handle_member_removal(self):
+        prev_doc = self.get_doc_before_save()
+        if not prev_doc:
+            return
+
+        current_user = frappe.session.user
+        current_session_data = frappe.session.data
+        frappe.set_user("Administrator")
```

Comment:
Tangential case: Can this be applied here too ? 

https://github.com/fossunited/fossunited/pull/525/files/8d9bfdfdc9f87e06571060bc5c610816101295c0#r1705559152

Reply:
I don't think `add_roles` & `remoge_role` work with `ignore_permission`

Reply:
I also have a doubt whether we need to give the Admin permissions here. 

The role addition is anyway executed in the Administrator level right? If it's server side code. We did the same for localhost role additions ? Afaik we did not make the user switch there ?

Reply:
No, the code is not executed on administrator level on server side. I tried it, it will throw a Permission error on User doctype.

Reply:
For the LocalHost, the System managers were the one who added the LocalHost members. System managers can make changes to the User doctype

Reply:
> No, the code is not executed on administrator level on server side. I tried it, it will throw a Permission error on User doctype.

That sort of reverts me back to the first doubt - can ignorePermissions not handle this?

Reply:
Again, as i already told, add_roles and remove_roles method does not work with ignore permissions. This is not a save operation or a create operation.

Reply:
As referred with Maharshi - the approach seems off, tho it seems our only visible option now. 

It would be good to see if there are alter options.

Reply:
Just add a tweaked version of utility function like following so that everyone is aware of what's going on instead of temporarily making a user Admin? 

with this
- You'll get the history of who made the changes if required
- And you don't have to worry about proper session restore

```python
def remove_user_roles(user_name, *roles):
	user = frappe.get_doc("User", user_name)
	existing_roles = {d.role: d for d in user.get("roles")}
	for role in roles:
		if role in existing_roles:
			user.get("roles").remove(existing_roles[role])
	user.save(ignore_permissions=True)
```

Reply:
brilliant @surajshetty3416 , this works!